### PR TITLE
add linux-syscall and linux-errno crate

### PR DESCRIPTION
**Pull request by haraldh**
includes no main Cargo.toml and Cargo.lock changes

### CODE REVIEW

**Comment by npmccallum** [comment]
@haraldh Just a design question: do we intend to support 32bit keeps on 64bit hosts or vice versa? I'm inclined to say no. Thoughts?

**Comment by haraldh** [comment]
added Cargo.toml ... otherwise travis is useless

**Comment by haraldh** [comment]
> @haraldh Just a design question: do we intend to support 32bit keeps on 64bit hosts or vice versa? I'm inclined to say no. Thoughts?

ah, right.. will fix

**Comment by haraldh** [comment]
> @haraldh Just a design question: do we intend to support 32bit keeps on 64bit hosts or vice versa? I'm inclined to say no. Thoughts?

no

**Comment by haraldh** [comment]
>     1. I don't see any reason to bundle errnos and syscalls. I can imagine people waning one without the other. Let's make these separate crates.
> 
>     2. We already have a convention for creating type-safe constants. See, for example: https://github.com/enarx/enarx/blob/master/sgx-types/src/sig.rs#L3-L13 I think both errnos and syscalls should follow this convention.

resolved both points

File: linux-errno/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,18 @@
+[package]
+name = "linux_errno"
```
[the remark this answers is not in the document]
Comment (haraldh) [reply linux-errno/Cargo.toml]: In the end, it's `_` ... Doesn't matter

File: linux-errno/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! errno type and constants for various architectures
+
+/// x86 errno
+pub mod x86_64_errno;
```
[the remark this answers is not in the document]
Comment (haraldh) [reply linux-errno/src/lib.rs]: Yeah

File: linux-errno/src/x86_64_errno.rs [hunk 3]
```diff
@@ -0,0 +1,460 @@
...
+impl Errno {
+    /// create a new instance
+    pub const fn new(id: i64) -> Self {
+        Self(id)
+    }
+
+    /// convert to the transparent type
+    pub const fn as_i64(&self) -> i64 {
+        self.0
+    }
```
[the remark this answers is not in the document]
Comment (haraldh) [reply linux-errno/src/x86_64_errno.rs]: match

File: linux-errno/src/x86_64_errno.rs [hunk 4]
```diff
@@ -0,0 +1,460 @@
...
+}
+
+impl From<i64> for Errno {
+    fn from(val: i64) -> Self {
+        Self::new(val)
+    }
+}
+
+impl Into<i64> for Errno {
```
[the remark this answers is not in the document]
Comment (haraldh) [reply linux-errno/src/x86_64_errno.rs]: Any reason?

File: linux-errno/src/x86_64_errno.rs [hunk 5]
```diff
@@ -0,0 +1,460 @@
...
+/// Errno as a transparent type
+#[repr(transparent)]
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub struct Errno(i64);
+
+impl Errno {
+    /// create a new instance
+    pub const fn new(id: i64) -> Self {
+        Self(id)
+    }
```
[the remark this answers is not in the document]
Comment (haraldh) [reply linux-errno/src/x86_64_errno.rs]: `match Errno::new(errno_last_error)`

File: linux-errno/src/x86_64_errno.rs [hunk 6]
```diff
@@ -0,0 +1,460 @@
...
+//   | egrep '^#define\s+E[A-Z_0-9]+\s+[0-9]+' \
+//   | while read a b c d; do \
+//       echo "$d" | sed -e 's#/\*#///#;s#\*/##'; \
+//       echo "pub const ${b} : Errno = Errno::new($c);"; \
+//       echo
+//     done
+// ```
+
+/// Operation not permitted
+pub const EPERM: Errno = Errno::new(1);
```
[the remark this answers is not in the document]
Comment (haraldh) [reply linux-errno/src/x86_64_errno.rs]: Ok

File: linux-syscall/src/x86_64_syscall.rs [hunk 7]
```diff
@@ -0,0 +1,1073 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! syscall type and constants
+
+/// SyscallNR as a transparent type
+#[repr(transparent)]
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub struct SyscallNR(u64);
```
[the remark this answers is not in the document]
Comment (haraldh) [reply linux-syscall/src/x86_64_syscall.rs]: The kernel constants are all named `__NR_…` and it's not a syscall, but only the number

[the remark this answers is not in the document]
Comment (haraldh) [reply linux-syscall/src/x86_64_syscall.rs]: And if imported with use, I want it to be distinct.

[the remark this answers is not in the document]
Comment (npmccallum) [reply linux-syscall/src/x86_64_syscall.rs]: `use syscall::Syscall as SyscallNum`

Stylistically, we don't care about reproducing c-isms.

File: linux-errno/src/x86_64_errno.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (npmccallum) [reply linux-errno/src/x86_64_errno.rs]: Ack

File: linux-errno/src/x86_64_errno.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (npmccallum) [reply linux-errno/src/x86_64_errno.rs]: I think it is more consistent to mark the `impl From` as inline. I think the end result would be the same but easier to read.

File: linux-errno/Cargo.toml (hunk 1 above)
[the remark this answers is not in the document]
Comment (npmccallum) [reply linux-errno/Cargo.toml]: It doesn't match our existing style.

File: linux-errno/src/x86_64_errno.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (npmccallum) [reply linux-errno/src/x86_64_errno.rs]: The reason for this is that `std` provides a default blanket implementation for `Into` for everything that implements `From`.

[the remark this answers is not in the document]
Comment (haraldh) [reply linux-errno/src/x86_64_errno.rs]: Yeah, yeah

[the remark this answers is not in the document]
Comment (haraldh) [reply linux-errno/src/x86_64_errno.rs]: Forgot, that you can impl for simple types

File: linux-errno/Cargo.toml (hunk 1 above)
[the remark this answers is not in the document]
Comment (steveeJ) [reply linux-errno/Cargo.toml]: TLDR: I'm in favor of `linux-errno`.

Reasoning:

1. Sticking to the existing style for consistency
1. From the [cargo reference on renaming dependencies](https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#renaming-dependencies-in-cargotoml):

    > When writing a [dependencies] section in Cargo.toml the key you write for a dependency typically matches up to the name of the crate you import from in the code. (...)
1. Consistency between the name for the package in the `members` array (which is given as the path relative to the workspace crate) with the package name

File: linux-syscall/src/x86_64_syscall.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (haraldh) [reply linux-syscall/src/x86_64_syscall.rs]: In that regard I would rather say:
`use syscall::Num as SyscallNum`

[the remark this answers is not in the document]
Comment (npmccallum) [reply linux-syscall/src/x86_64_syscall.rs]: WFM

File: linux-errno/Cargo.toml (hunk 1 above)
[the remark this answers is not in the document]
Comment (haraldh) [reply linux-errno/Cargo.toml]: fixed

File: linux-errno/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+
+#[cfg(target_arch = "x86_64")]
+pub use x86_64_errno as errno;
```
[the remark this answers is not in the document]
Comment (haraldh) [reply linux-errno/src/lib.rs]: fixed

File: linux-errno/src/x86_64_errno.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (haraldh) [reply linux-errno/src/x86_64_errno.rs]: fixed

File: linux-syscall/src/x86_64_syscall.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (haraldh) [reply linux-syscall/src/x86_64_syscall.rs]: fixed

File: linux-errno/src/x86_64_errno.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (haraldh) [reply linux-errno/src/x86_64_errno.rs]: fixed

File: linux-errno/src/lib.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (haraldh) [reply linux-errno/src/lib.rs]: fixed

File: linux-errno/src/x86_64_errno.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (haraldh) [reply linux-errno/src/x86_64_errno.rs]: fixed

**Comment by haraldh** [comment]
> This all looks good to me. Unfortunately, GitHub won't let me merge without a manual rebase. Can you rebase on the latest master?

done

**Comment by haraldh** [comment]
ready to merge

**Comment by haraldh** [comment]
oops... forgot `no_std` feature

**Comment by npmccallum** [comment]
> oops... forgot `no_std` feature

Why do we need a `no_std` feature? Can't the libraries just be unconditionally `no_std`?

**Comment by haraldh** [comment]
> > oops... forgot `no_std` feature
> 
> Why do we need a `no_std` feature? Can't the libraries just be unconditionally `no_std`?

no tests are run if `no_std` is enabled

**Comment by haraldh** [comment]
> > > oops... forgot `no_std` feature
> > 
> > 
> > Why do we need a `no_std` feature? Can't the libraries just be unconditionally `no_std`?
> 
> no tests are run if `no_std` is enabled

oops.. pebkac ... of course it's possible... removed the features and made them `#![no_std]`

**Comment by npmccallum** [comment]
@haraldh Congratulations on your first merge!

**Pull request merged**